designdocs/recovery_system.md

```diff
@@ -0,0 +1,609 @@
+
+// TODO: Explore efficiency for this storage and if we can avoid double indexing
+/// Storage type for a Recovery Commitment to MsaId mapping
+/// - Key: RecoveryCommitment
+/// - Value: MsaId
+#[pallet::storage]
+pub type RecoveryCommitmentToMsaId<T: Config> = StorageMap<_, Twox64Concat, RecoveryCommitment, MsaId>;
+```
+
+## Required Extrinsics
+
+|Name/Description|Caller|Payment|Key Events|Runtime Added|
+|---|---|---|---|---|
+| `add_recovery_commitment`<br />Add a new Recovery Commitment to an existing MSA| Provider| Capacity or Tokens | [`RecoveryCommitmentAdded`](https://frequency-chain.github.io/frequency/pallet_msa/pallet/enum.Event.html#variant.RecoveryCommitmentAdded)| 166|
+| `verify_intermediary_hashes`<br />Verify Recovery Intermediary Hashes against stored Recovery Commitment| Recovery Provider| Capacity or Tokens | [`RecoveryCommitmentVerified`](https://frequency-chain.github.io/frequency/pallet_msa/pallet/enum.Event.html#variant.RecoveryCommitmentVerified)| 166|
+| `recover_account`<br />Recover MSA with new control key| Recovery Provider | Capacity or Tokens | [`AccountRecovered`](https://frequency-chain.github.io/frequency/pallet_msa/pallet/enum.Event.html#variant.AccountRecovered), [`RecoveryCommitmentInvalidated`](https://frequency-chain.github.io/frequency/pallet_msa/pallet/enum.Event.html#variant.RecoveryCommitmentInvalidated) | 166|
+
+### Add Recovery Commitment
+
+This extrinsic allows a provider to add a new Recovery Commitment to an existing MSA. It can be used during account creation or later to update the Recovery Commitment.
+
+```rust
+// Pseudo-code for adding a Recovery Commitment to an existing MSA
+pub fn add_recovery_commitment(
+    origin: OriginFor<T>,
+    msa_owner_key: T::AccountId,
+    proof: MultiSignature,
+    payload: RecoveryCommitmentPayload,
+) -> DispatchResult {
+    let provider_key = ensure_signed(origin)?;
+
+    // Verify that the MsaId owner has signed the payload
+    ensure!(
+        Self::verify_signature(&proof, &msa_owner_key, &payload),
+        Error::<T>::InvalidSignature
+    )
+
+    // Register the signature to prevent replay attacks
+    Self::register_signature(&proof, payload.expiration.into());
+    
+    // Recover the MSA ID from the msa_owner_key
+    let msa_id = ensure_valid_msa_key(&msa_owner_key);
+
+    // Store the new RecoveryCommitment
+    MsaIdToRecoveryCommitment::<T>::insert(msa_id, payload.recovery_commitment);
+    Self::deposit_event(Event::RecoveryCommitmentAdded {
+      who: msa_owner_key,
+      msa_id,
+    });
+
+    Ok(())
+}
+```
+
+### Verify Intermediary Hashes
+
+This extrinsic is used by the Recovery Provider to verify if a user’s Recovery Intermediary Hashes compute a match to any existing MSA.
+
+```rust
+// Pseudo-code for verifying the Recovery Intermediary Hashes
+pub fn verify_intermediary_hashes(
+    payload: IntermediaryHashesPayload,
```

Comment:
I'm tempted to say we should just do two parameters instead of a payload system as it doesn't have an external signer.